db/memtable.cc

```diff
@@ -532,38 +532,272 @@ Status MemTable::VerifyEncodedEntry(Slice encoded,
+      if (end_compare < 0) {
+        auto k = std::make_pair(key, value);
+        to_insert[k].insert(tombstone_seqs.begin(), tombstone_seqs.end());
+        to_insert[k].insert(s);
+        to_insert[std::make_pair(value, tombstone_end_key)].insert(
+            tombstone_seqs.begin(), tombstone_seqs.end());
+      } else if (end_compare == 0) {
+        auto k = std::make_pair(key, value);
+        to_insert[k].insert(tombstone_seqs.begin(), tombstone_seqs.end());
+        to_insert[k].insert(s);
+      } else if (user_comparator->Compare(key, tombstone_end_key) < 0) {
+        auto k = std::make_pair(key, tombstone_end_key);
+        to_insert[k].insert(tombstone_seqs.begin(), tombstone_seqs.end());
+        to_insert[k].insert(s);
+        to_insert[std::make_pair(tombstone_end_key, value)].insert(s);
+      } else {
+        to_insert[std::make_pair(key, value)].insert(s);
+      }
+    } else if (start_compare < 0) {
+      int end_compare = user_comparator->Compare(value, tombstone_end_key);
+      if (end_compare > 0) {
+        to_insert[std::make_pair(key, tombstone_start_key)].insert(s);
+        auto k = std::make_pair(tombstone_start_key, tombstone_end_key);
+        to_insert[k].insert(tombstone_seqs.begin(), tombstone_seqs.end());
+        to_insert[k].insert(s);
+        to_insert[std::make_pair(tombstone_end_key, value)].insert(s);
+      } else if (user_comparator->Compare(value, tombstone_start_key) > 0) {
+        to_insert[std::make_pair(key, tombstone_start_key)].insert(s);
+        auto k = std::make_pair(tombstone_start_key, value);
+        to_insert[k].insert(tombstone_seqs.begin(), tombstone_seqs.end());
+        to_insert[k].insert(s);
+        if (end_compare < 0) {
+          to_insert[std::make_pair(value, tombstone_end_key)].insert(
+              tombstone_seqs.begin(), tombstone_seqs.end());
+        }
+      } else {
+        to_insert[std::make_pair(key, value)].insert(s);
+      }
+    } else {
+      int end_compare = user_comparator->Compare(value, tombstone_end_key);
+      if (end_compare < 0) {
+        auto k = std::make_pair(key, value);
+        to_insert[k].insert(tombstone_seqs.begin(), tombstone_seqs.end());
+        to_insert[k].insert(s);
+        to_insert[std::make_pair(value, tombstone_end_key)].insert(
+            tombstone_seqs.begin(), tombstone_seqs.end());
+      } else if (end_compare > 0) {
+        auto k = std::make_pair(key, tombstone_end_key);
+        to_insert[k].insert(tombstone_seqs.begin(), tombstone_seqs.end());
+        to_insert[k].insert(s);
+        to_insert[std::make_pair(tombstone_end_key, value)].insert(s);
+      } else {
+        auto k = std::make_pair(key, value);
+        to_insert[k].insert(tombstone_seqs.begin(), tombstone_seqs.end());
+        to_insert[k].insert(s);
+      }
+    }
+  }
+
+  if (to_insert.empty()) {
+    to_insert[std::make_pair(key, value)].insert(s);
+  }
+  uint32_t entries = 0;
+  uint32_t len = 0;
+  for (auto& i : to_insert) {
```

Comment:
Depending on the order of insertion, there may be a potential race condition with reading:

current range tombstone memtable has `[a, d)@1`
concurrent range delete: `[c, f)@2`
to_insert: `[c, d)@[2,1], [d,f)@2`
concurrent read: key `e` at snapshot 1 


If max covering tombstone seq is computed after `[c,d)@[2,1]` is inserted, but before `[d,f)@2` is inserted, then we might incorrectly conclude that there is no covering tombstone?

Reply:
I think when finding e@1, [d,f)@2 should not be found, so no covering tombstone is the expected state?

Reply:
Ah, yes, that's a bad example. Let's consider this one:

memtable has: `[a, f)@2`
current range delete: `[c, d)@3`
to_insert: `[a, c)@2`, `[c, d)@[3, 2]`, `[d, f)@2`
current read: key `e` at snapshot 2.

The expected max covering tombstone seq is 2 from `[a, f)@2`. If max covering seqno is computed after `[c, d)@[3, 2]` is inserted but before `[d, f)@2`, then we might find there is no covering range tombstone?